### ISSUE TITLE
--preserve-full-path does not generate the full .h path in #include

### ISSUE THREAD

**Issue by priand** [google/j2objc#1030]
See the discussion here: https://groups.google.com/forum/#!searchin/j2objc-discuss/riand%7Csort:date/j2objc-discuss/_fIt8pTKdCw/qGkfA09SDAAJ

When the outputStyle is set to SOURCE, it should still generate full path for the included header files, as the files are generated hierarchically. Currently, it only generates the full path when the option is PACKAGE.
I fixed the issue in `HeaderMap.outpDirFromPackage`, by simply adding the `case SOURCE:`

```
  private String outputDirFromPackage(PackageElement pkg) {
    if (pkg == null || pkg.isUnnamed()) {
      return "";
    }
    String pkgName = ElementUtil.getName(pkg);
    OutputStyleOption style = outputStyle;
    if (isPlatformPackage(pkgName)) {
      // Use package directories for platform classes if they do not have an entry in the header
      // mapping.
      style = OutputStyleOption.PACKAGE;
    }
    switch (style) {
      case SOURCE:
      case PACKAGE:
        return ElementUtil.getName(pkg).replace('.', File.separatorChar) + File.separatorChar;
      default:
        return "";
    }
  }
```

**Comment by tomball** [comment]
This change breaks three unit tests, suggesting that SOURCE isn't the solution you are looking for. The SOURCE output is for projects that need to build using the relative path of a file, regardless of its Java package. This can reduce the number of compiler include paths, but at the expense of harder to read generated code. Very few projects are big enough to justify that trade-off.

Is your project trying to transpile JRE or Guava classes? [HeaderMap has a list of package roots](https://github.com/google/j2objc/blob/master/translator/src/main/java/com/google/devtools/j2objc/util/HeaderMap.java#L46) that always have package directories. Perhaps that's what you are running into.

Regardless, it sounds like your project is too complex to use the --no-package-directories flag. That flag was added to support beginners whose main (sometimes only) IDE experience was Xcode, and for whom adding an external build rule that invokes something like make or Gradle was beyond their ability. If you translate your Java code using an external build rule, then the default package output style will prevent any class name collisions.

**Comment by priand** [comment]
Hi Tom, thanks for your quick reply. Maybe I misinterpreted what `--preserve-full-paths` is for?

The `--preserve-full-paths` property generates the right directory structure, with the .h and .m in their respective package directories, but the generated`#include` do not use the full path, only the class name. This is because `--preserve-full-paths` sets the style options to SOURCE which, according to the code above, doesn't use the full path for includes. It seems to be an inconsistency to me: why would you generate the .h into sub directories and not include the sub directory in the file names you import?

But my actual issue might also be related to the maven plug-in. When the `preserveFullPath` option is not set, it generates the source files of the project and its dependencies like bellow:
```
/<project classes>
/<dep1>/<dep1 classes>
/<dep2>/<dep2 classes>
```
which is hard to exploit from XCode. I would like all the classes in the same folder hierarchy, which is what it does when the flag is set:
```
/<project and dependency classes in the same hierarchy>
```

The headerMap is an idea, although it is not currently configurable with our custom package names. BTW, are all the com.google.common.* worth in the PLATFORM_PACKAGES set when com.google.common is added?

**Comment by tomball** [comment]
This should be fixed by [f10be7d ](https://github.com/google/j2objc/commit/f10be7d60c0deabe57af8903b112e03c972229a1). Please reopen if that's not the case.
